lib/acts_as_span/end_date_propagator.rb

```diff
@@ -0,0 +1,111 @@
+# End Date Propagator
+#
+# When editing the `end_date` of a record, the record's children often also
+#   need to be updated. This propagator takes care of that.
+# For each of the  child records (defined below in the function `children`),
+#   the child record's `end_date` is updated to match that of the original
+#   object. The main function `call` is recursive, propagating to children of
+#   children and so on.
+# Records that should not have their end dates propagated in this manner
+#   (e.g. StatusRecords) are manually excluded in `non_ended_classes`.
+# If there is some error preventing propagation, the child record is NOT saved,
+#   and that error message is added to the object's `errors`. These errors
+#   propagate upwards into a flattened array of error messages.
+#
+# Currently only propagates "default" span. The approach to implementing such
+#   a feature is ambiguous - would all children have the same span propagated?
+#   Would each acts_as_span model need a method to tell which span to
+#   propagate to? Once there is a solid use case for using this object on
+#   models with multiple spans, that will inform the implementation strategy.
+module ActsAsSpan
+  class EndDatePropagator
+    attr_reader :object,
+      :errors_cache,
+      :skipped_classes
+
+    def initialize(object, skipped_classes = [])
+      @object = object
+      @errors_cache = []
+      @skipped_classes = skipped_classes
+    end
+
+    def call
+      # no propagation if no span to propagate to
+      return object unless object.respond_to?(:span)
+      # no propagation if trying to propagate nil
+      unless end_date_changed?(object) && !object.span.end_date.nil?
+        return object
+      end
+
+      children(object).each do |child|
+        # skip child if child end_date is greater than parent end_date
+        next unless child.span.end_date.nil? ||
+          child.span.end_date > object.span.end_date
+
+        child.assign_attributes(
+          { child.span.end_field => object.span.end_date }
+        )
+
+        # End the record, its children too. And their children, forever, true.
+        child_obj = ActsAsSpan::EndDatePropagator.new(
```

Comment:
I got confused between child and child_obj
```suggestion
        child_propagator = ActsAsSpan::EndDatePropagator.new(
```

Reply:
this would be something more like `propagated_child `... the distinction is that it may have had errors added to it via the propagator